node/src/effect.rs

```diff
@@ -1532,8 +1532,40 @@ impl<REv> EffectBuilder<REv> {
                     key_block
                         .header()
                         .next_era_validator_weights()
-                        .map(|validators| validators.contains_key(validator))
+                        .map(|validators| validators.contains_key(&validator))
                 })
+        };
+
+        match maybe_is_currently_bonded {
+            // if we know whether the validator is bonded, just return that
+            Some(is_bonded) => Ok(is_bonded),
+            // if not, try checking future eras with the latest state root hash
+            None => match latest_state_root_hash {
+                // no root hash later than initial -> we just assume the validator is not bonded
+                None => Ok(false),
+                // otherwise, check with contract runtime
+                Some(state_root_hash) => self
+                    .make_request(
+                        |responder| ContractRuntimeRequest::IsBonded {
+                            state_root_hash,
+                            era_id,
+                            protocol_version,
+                            public_key: validator,
+                            responder,
+                        },
+                        QueueKind::Regular,
+                    )
+                    .await
+                    .or_else(|error| {
+                        // Promote this error to a non-error case.
+                        // It's not an error that we can't find the era that was requested.
+                        if error.is_era_validators_missing() {
+                            Ok(false)
+                        } else {
+                            Err(error)
+                        }
+                    }),
+            },
```

Comment:
(I would do _so_ many early returns in this method… but I know @goral09 doesn't like them.)

Reply:
If it improves readability by a lot then I am fine with them.

Reply:
I agree.  The cyclomatic complexity here is pretty high, and it's hard (for me) to reason about.